### ISSUE TITLE
podman hangs with deadlock

### ISSUE THREAD

**Issue by unxfrek** [containers/libpod#1574]
**Is this a BUG REPORT or FEATURE REQUEST?**:

[//]: # Uncomment only one, leave it on its own line:

 **kind bug**

[//]: # **kind feature**

**Description**

podman deadlocks on /var/lib/containers/storage/storage.lock

```
-bash-4.2$ ps -ef | grep podman
username   5343 29530  0 20:23 pts/1    00:00:00 grep --color=auto podman
root     13804     1  0 Sep27 ?        00:00:08 /usr/bin/podman --root /var/lib/containers/storage --runroot /var/run/containers/storage --log-level error --cgroup-manager cgroupfs --tmpdir /var/run/libpod --storage-driver overlay container cleanup 62bedf7f6d3a55020e7f314f401769281ef5b16c222d1753b0281590b48f63bb
root     14038 13701  0 Sep27 ?        00:02:25 /usr/bin/podman wait container_1537988606812_0012_01_000005
root     15355     1  0 Sep27 ?        00:00:07 /usr/bin/podman --root /var/lib/containers/storage --runroot /var/run/containers/storage --log-level error --cgroup-manager cgroupfs --tmpdir /var/run/libpod --storage-driver overlay container cleanup cf2ac277eb9a5b85e1b78ef088a5fff8ae41ac29639d27f77ade427211378d87
root     15359     1  0 Sep27 ?        00:00:07 /usr/bin/podman --root /var/lib/containers/storage --runroot /var/run/containers/storage --log-level error --cgroup-manager cgroupfs --tmpdir /var/run/libpod --storage-driver overlay container cleanup 79181a9932ae72ae4fe7b8377ab92cc57d1651dcd7e47dca4596b29b3787e934
root     15543 15298  0 Sep27 ?        00:01:51 /usr/bin/podman wait container_1537988606812_0012_01_000011
root     15546     1  0 Sep27 ?        00:00:08 /usr/bin/podman --root /var/lib/containers/storage --runroot /var/run/containers/storage --log-level error --cgroup-manager cgroupfs --tmpdir /var/run/libpod --storage-driver overlay container cleanup eb569521b8838b1336e6a23428c3f033ee981f7113d5ccc50967f0775dfae5dc
root     15564 15297  0 Sep27 ?        00:01:56 /usr/bin/podman wait container_1537988606812_0012_01_000017
root     15769 15404  0 Sep27 ?        00:01:56 /usr/bin/podman wait container_1537988606812_0012_01_000016
root     22399  9698  0 08:01 ?        00:00:00 [podman] <defunct>
```

**Steps to reproduce the issue:**

1.  Launch several containers at the same time

2. wait for pod man to hang

**Output of `podman info`:**

```
$ sudo podman info
host:
  Conmon:
    package: podman-0.9.4-1537914470.gitf4e2810f.el7.x86_64
    path: /usr/libexec/podman/conmon
    version: 'conmon version 1.12.0-dev, commit: 51dcc3ad101da1c64efd9d49a1a0911242beac29-dirty'
  MemFree: 9827209216
  MemTotal: 16557703168
  OCIRuntime:
    package: runc-1.0.0-37.rc5.dev.gitad0f525.el7.x86_64
    path: /usr/bin/runc
    version: 'runc version spec: 1.0.0'
  SwapFree: 0
  SwapTotal: 0
  arch: amd64
  cpus: 4
  hostname: host01.example.com
  kernel: 3.10.0-862.9.1.el7.x86_64
  os: linux
  uptime: 702h 19m 22.38s (Approximately 29.25 days)
insecure registries:
  registries: []
registries:
  registries:
  - registry.access.redhat.com
store:
  ContainerStore:
    number: 5
  GraphDriverName: overlay
  GraphOptions:
  - overlay.override_kernel_check=true
  GraphRoot: /var/lib/containers/storage
  GraphStatus:
    Backing Filesystem: xfs
    Native Overlay Diff: "true"
    Supports d_type: "true"
  ImageStore:
    number: 2
  RunRoot: /var/run/containers/storage
```
Attaching a file with the strace and pstack since it is large.

[podman_deadlock_modified.txt](https://github.com/containers/libpod/files/2435498/podman_deadlock_modified.txt)

**Comment by mheon** [comment]
Can you get the output of `lslocks` during the deadlock?

**Comment by unxfrek** [comment]
```
-bash-4.2$ lslocks
COMMAND           PID   TYPE SIZE MODE   M      START        END PATH
lvmetad           460  POSIX   0B WRITE  0          0          0 /run/netns
osqueryd         1474  POSIX   0B WRITE  0          0          0 /var/lib/containers/storage/overlay
osqueryd         1474  FLOCK   0B WRITE  0          0          0 /var/lib/containers/storage/overlay
crond            2242  FLOCK   0B WRITE  0          0          0 /run/netns
java            12368  POSIX   0B WRITE  0          0          0 /var/lib/containers/storage/overlay
nslcd            1327  POSIX   0B WRITE  0          0          0 /run/netns
dockerd-current  2330  FLOCK   0B WRITE  0          0          0 /var/lib/containers/storage/overlay
nessusd         31866  POSIX   0B READ   0 1073741826 1073742335 /var/lib/containers/storage/overlay
nessusd         31866  POSIX   0B READ   0        128        128 /var/lib/containers/storage/overlay
java            12368  POSIX   0B WRITE  0          0          0 /var/lib/containers/storage/overlay
java            12368  POSIX   0B WRITE  0          0          0 /var/lib/containers/storage/overlay
yakl-client     10832  FLOCK   0B WRITE  0          0          0 /var/lib/containers/storage/overlay
falcond           655  POSIX   0B WRITE  0          0          0 /run/netns
master           1905  FLOCK   0B WRITE  0          0          0 /var/lib/containers/storage/overlay
master           1905  FLOCK   0B WRITE  0          0          0 /var/lib/containers/storage/overlay
(unknown)          -1 OFDLCK   0B READ   0          0          0 /dev
supervise        9316  FLOCK   0B WRITE  0          0          0 /var/lib/containers/storage/overlay
ykeykeyd         9317  POSIX   0B WRITE  0          0          0 /var/lib/containers/storage/overlay
jsvc            16188  POSIX   0B WRITE  0          0          0 /var/lib/containers/storage/overlay
java            12368  POSIX   0B WRITE  0          0          0 /var/lib/containers/storage/overlay
java            12368  POSIX   0B WRITE  0          0          0 /var/lib/containers/storage/overlay
podman          32595  POSIX   0B WRITE* 0          0          0 /run/netns
podman          32736  POSIX   0B WRITE  0          0          0 /run/netns
atd              2247  POSIX   0B WRITE  0          0          0 /run/netns
supervise        9315  FLOCK   0B WRITE  0          0          0 /var/lib/containers/storage/overlay
multilog         9318  FLOCK   0B WRITE  0          0          0 /var/lib/containers/storage/overlay
```

**Comment by mheon** [comment]
Hm. It looks like Nessus has grabbed a POSIX lock on /var/lib/containers/storage/overlay. C/storage uses flock, but it's possible it's still interfering with Podman.

**Comment by unxfrek** [comment]
The way I get it to deadlock quickly is by running a Hadoop job that spawns JVMs that touch a lot of memory and try to make the VM swap. This is a test that I have been running on docker to get docker to fail with the containerd vs dockerd out of sync issue.

**Comment by mheon** [comment]
Hm. The locks in question aren't mandatory, though.

One Podman is stuck waiting for /var/run/netns, which sounds like the CNI plugins?

**Comment by mheon** [comment]
@unxfrek Are you spawning the JVMs in containers, or are they just a way to make the host run out of memory and force swapping?

**Comment by baude** [comment]
eliminate the networking component with --network=none ?

**Comment by TomSweeneyRedHat** [comment]
@unxfrek any chance you could share the Hadoop job or otherwise provide a reproducer?  It would be handy for our diagnosis and future testings.

**Comment by unxfrek** [comment]
They've been able to turn off nessus on the VMs that are being used to test podman and rerun the tests. However, they still are able to generate a deadlock on each VM running the same test. I've attached logs for all 3 of these VMs. Two things are particularly interesting:

1. The deadlock only happens on `podman ps -a`. I can still run containers via `podman run` perfectly fine. This is an upgrade over the previous deadlocks that I was seeing where everything related to podman would hang.

2. The straces of the `podman ps -a` on each VM seem to stick at different points in the code. But for each VM, they always stick during those same pieces of the code across different runs of `podman ps -a`

I'll see about a reproducer.

- Jacob
[podman_deadlock_no_nessus_10_5.3.txt](https://github.com/containers/libpod/files/2460402/podman_deadlock_no_nessus_10_5.3.txt)
[podman_deadlock_no_nessus_10_5.2.txt](https://github.com/containers/libpod/files/2460403/podman_deadlock_no_nessus_10_5.2.txt)
[podman_deadlock_no_nessus_10_5.1.txt](https://github.com/containers/libpod/files/2460404/podman_deadlock_no_nessus_10_5.1.txt)

**Comment by mheon** [comment]
@ericbadger Thanks - that's very thorough!

This is seeming like `runc state` getting stuck somewhere and not returning, causing us to get caught in a critical section.

We can fix this locally by adding a timeout on our call to runc - 10 seconds, maybe? - after which we kill runc and abandon the operation.

That's just treating the symptoms, though. I'm betting the root cause here is a runc bug causing it to occasionally hang when doing `runc status` calls. Can you get a more detailed `ps` of the `runc` processes (something that shows process state, `ps aux` should be fine) and maybe try and strace, see if it's trying to grab a lock or similar?

**Comment by ericbadger** [comment]
@mheon sorry for not including the strace/pstack in the last comment. I had the output, just forgot to include it in the comment. Anyway, it's stuck waiting on some futex. Not sure where to go from there on the debugging front.

```
[root@hostname1 overlay-containers]# ps aux | grep runc
root      1135  0.0  0.0  83936  2404 ?        S    Oct01   0:00 /usr/bin/runc state 823c299634ab27b42936a5edc84ec7cba084924f98c0ff69c26a9d0a26738041
root      1752  0.0  0.0  83936  2404 ?        S    Oct01   0:00 /usr/bin/runc state d292af5d6733789beb8ea50a49ba6fb2bea74667071cd920f2459710c506bd17
root      2330  0.1  0.1 1007288 26992 ?       Ssl  Aug30 102:05 /usr/bin/dockerd-current --add-runtime docker-runc=/usr/libexec/docker/docker-runc-current --default-runtime=docker-runc --authorization-plugin=rhel-push-plugin --exec-opt native.cgroupdriver=cgroupfs --userland-proxy-path=/usr/libexec/docker/docker-proxy-current --selinux-enabled -g /home/gs/var/docker --add-registry docker-registry.ops.yahoo.com:4443 --add-registry registry.access.redhat.com
root      2517  0.1  0.0 769096  8328 ?        Ssl  Aug30  62:17 /usr/bin/docker-containerd-current -l unix:///var/run/docker/libcontainerd/docker-containerd.sock --metrics-interval=0 --start-timeout 2m --state-dir /var/run/docker/libcontainerd/containerd --shim docker-containerd-shim --runtime docker-runc
root      5784  0.0  0.0  83936  2408 ?        S    Oct05   0:00 /usr/bin/runc state 039c73bae701c7fc6ecd3dfe2b52175f5c4e1406dc1b98677078e2a0218e5fe4
root      6648  0.0  0.0  83936  2408 ?        S    Oct05   0:00 /usr/bin/runc state 1b62a8299325a002b51c3d081f9d7daa3c23526837b8cfa9f2f4d08f70639e1f
root      7808  0.0  0.0 112708   980 pts/0    S+   19:58   0:00 grep --color=auto runc
root     15957  0.0  0.0  83936  2404 ?        S    Sep27   0:00 /usr/bin/runc state 62bedf7f6d3a55020e7f314f401769281ef5b16c222d1753b0281590b48f63bb
root     15970  0.0  0.0  83936  2404 ?        S    Sep27   0:00 /usr/bin/runc state cf2ac277eb9a5b85e1b78ef088a5fff8ae41ac29639d27f77ade427211378d87
root     15971  0.0  0.0  83936  2404 ?        S    Sep27   0:00 /usr/bin/runc state 79181a9932ae72ae4fe7b8377ab92cc57d1651dcd7e47dca4596b29b3787e934
root     15973  0.0  0.0  83936  2404 ?        S    Sep27   0:00 /usr/bin/runc state eb569521b8838b1336e6a23428c3f033ee981f7113d5ccc50967f0775dfae5dc
root     23669  0.0  0.0  83936  2408 ?        S    Oct03   0:00 /usr/bin/runc state b82400095591ef20172eab9d5842de881b0798aa5c9bc8c84338779e8a0d4c5f
```

```
[root@hostname1 overlay-containers]# strace -fp 5784
strace: Process 5784 attached
futex(0x562902da4360, FUTEX_WAIT, 2, NULL^Cstrace: Process 5784 detached
 <detached ...>
```

```
[root@hostname1 overlay-containers]# pstack 5784
#0  0x000056290261e1c3 in runtime.futex ()
#1  0x00005629025eb842 in runtime.futexsleep ()
#2  0x0000562902da4360 in __bss_start ()
#3  0x0000000200000000 in ?? ()
#4  0x0000000000000000 in ?? ()
```

**Comment by rhatdan** [comment]
@mheon why do we need to run runc status?  Is there some piece of data that we need from runc that could cause this issue?  Could their be a runc on a container as it exits that causes this?
I think a timeout of 1 second would be better then waiting 10 seconds.

**Comment by mheon** [comment]
@rhatdan We need to ping runc to see if the container is alive or dead, so I don't think we can get away with getting rid of it (though moving to cleanup processes for DB updates would substantially reduce the amount of them we need)

**Comment by mheon** [comment]
I think this is a runc bug of some sort, though. It seems to be stuck in a futex, so it could be a race condition

**Comment by rhatdan** [comment]
Can't we rely on conmon to have told us that by dropping an exit code?  If their is no exit code then we figure the container is alive.  Maybe we add a flag to podman ps -a --fullcheck or something that will check with the container runtime.

**Comment by ericbadger** [comment]
Is there any additional information or testing that I can do to help out with this? I'd really like to get to the bottom of what's going on here so that we can deploy podman at scale. I just don't know the internals of podman or runc :)

**Comment by mheon** [comment]
I'll see about getting a temporary fix in for this this week, using a timeout on runc.

I'm still confident the underlying cause is a runc bug, though not exactly sure where. @mrunalp Anything debugging-wise we can do to chase this down?

**Comment by mrunalp** [comment]
@mheon Try stracing runc to see where it is hanging. Also, like @rhatdan said, you shouldn't have to call runc state frequently. CRI-O uses inotify on conmon created exit file to know when a container has stopped.

**Comment by mheon** [comment]
Part of me wonders if this isn't related to https://github.com/containers/libpod/pull/1681 - we could be seeing a lot of defunct runc processes accumulating without that PR, especially with wait

**Comment by ericbadger** [comment]
@mrunalp I posted the strace of the runc command that is hung 
 in [this](https://github.com/containers/libpod/issues/1574#issuecomment-428331006) comment above

**Comment by mheon** [comment]
Initial work on resolving this in #1689

**Comment by mheon** [comment]
Further update: I'm confident #1689 is mostly ready, but it's causing strange failures in our CI. I've spent several days tracking these down, so far no luck.

**Comment by vrothberg** [comment]
@mheon, has this issue been resolved in the meantime?

**Comment by mheon** [comment]
Yeah, this should be fixed via #1689 - if it's not, feel free to reopen
